### PR TITLE
More cmsg cleanup

### PR DESCRIPTION
**Pull request by Ralith**
No semantic changes, just additional insurance against UB.

### CODE REVIEW

File: quinn/src/platform/cmsg.rs [hunk 1]
```diff
@@ -34,8 +42,10 @@ impl<'a> Encoder<'a> {
         self.len += space;
         self.cmsg = unsafe { libc::CMSG_NXTHDR(self.hdr, cmsg).as_mut() };
     }
+}
 
-    pub fn finish(self) {
+impl<'a> Drop for Encoder<'a> {
```
Comment (djc) [inline quinn/src/platform/cmsg.rs]: Do you really feel this is better than what you had before? It feels less idiomatic to me.

Comment (Ralith) [reply quinn/src/platform/cmsg.rs]: I was concerned that the prior approach was easy to forget to do. With the new approach, borrowck forces you to ensure the `Encoder` is dropped before the buffer is used. I'm open to other ideas, though.

Comment (djc) [reply quinn/src/platform/cmsg.rs]: I understand the concern. I think the way I would write it is to have `finish()` return an opaque newtype around the buffer which is what's actually passed to other code. Would that make sense here?

Comment (Ralith) [reply quinn/src/platform/cmsg.rs]: The "other code" in this case is libc, so there's not an opportunity for a richer interface there.

Comment (Ralith) [reply quinn/src/platform/cmsg.rs]: Hmm, maybe `Encoder` could take ownership of the buffer and only release it on `finish`.

Comment (Ralith) [reply quinn/src/platform/cmsg.rs]: No, that won't work, it needs a stable address to be referenced in the `msghdr`.

**Comment by Ralith** [comment]
Switched to a compromise approach which allows explicit finishing of the encoding operation while preserving the static guarantee.

**Comment by djc** [comment]
Looks okay to me once CI passes.

**Comment by Ralith** [comment]
Should be ready.

**Pull request merged**